generator/altitude_generator.cpp

```diff
@@ -0,0 +1,247 @@
+      }
+      if (!allPointsValidAltFlag)
+        break;
+
+      hasAltitude = true;
+      m_featureAltitudes.push_back(make_pair(id, Altitude(move(altitudes))));
+
+      if (m_minAltitude == kInvalidAltitude)
+        m_minAltitude = minFeatureAltitude;
+      else
+        m_minAltitude = min(minFeatureAltitude, m_minAltitude);
+    }
+    while(false);
+
+    m_altitudeAvailability.push_back(hasAltitude);
+  }
+
+  bool HasAltitudeInfo()
+  {
+    return !m_featureAltitudes.empty();
+  }
+
+  void SortFeatureAltitudes()
+  {
+    sort(m_featureAltitudes.begin(), m_featureAltitudes.end(), my::LessBy(&Processor::TFeatureAltitude::first));
+  }
+
+private:
+  generator::SrtmTileManager m_srtmManager;
+  TFeatureAltitudes m_featureAltitudes;
+  vector<bool> m_altitudeAvailability;
+  TAltitude m_minAltitude;
+};
+
+uint32_t GetFileSize(string const & filePath)
+{
+  uint64_t size;
+  if (!my::GetFileSize(filePath, size))
+  {
+    LOG(LWARNING, ("altitudeAvailability", filePath, "size = 0"));
+    return 0;
+  }
+
+  LOG(LINFO, ("altitudeAvailability ", filePath, "size =", size, "bytes"));
+  return size;
+}
+
+void MoveFileToAltitudeSection(string const & filePath, uint32_t fileSize, FileWriter & w)
+{
+  {
+    ReaderSource<FileReader> r = FileReader(filePath);
+    w.Write(&fileSize, sizeof(fileSize));
+    rw::ReadAndWrite(r, w);
+    LOG(LINFO, (filePath, "size is", fileSize));
+  }
+  FileWriter::DeleteFileX(filePath);
+}
+
+void SerializeHeader(TAltitudeSectionVersion version, TAltitude minAltitude,
+                     TAltitudeSectionOffset altitudeInfoOffset, FileWriter & w)
+{
+  w.Write(&version, sizeof(version));
+
+  w.Write(&minAltitude, sizeof(minAltitude));
+  LOG(LINFO, ("altitudeAvailability writing minAltitude =", minAltitude));
+
+  w.Write(&altitudeInfoOffset, sizeof(altitudeInfoOffset));
+  LOG(LINFO, ("altitudeAvailability writing altitudeInfoOffset =", altitudeInfoOffset));
+}
+}  // namespace
+
+namespace routing
+{
+void BuildRoadAltitudes(string const & srtmPath, string const & baseDir, string const & countryName)
+{
+  try
+  {
+    // Preparing altitude information.
+    LOG(LINFO, ("srtmPath =", srtmPath, "baseDir =", baseDir, "countryName =", countryName));
+    string const mwmPath = my::JoinFoldersToPath(baseDir, countryName + DATA_FILE_EXTENSION);
+
+    Processor processor(srtmPath);
+    feature::ForEachFromDat(mwmPath, processor);
+    processor.SortFeatureAltitudes();
+    Processor::TFeatureAltitudes const & featureAltitudes = processor.GetFeatureAltitudes();
+
+    if (!processor.HasAltitudeInfo())
+    {
+      LOG(LINFO, ("No altitude information for road features of mwm", countryName));
+      return;
+    }
+
+    // Writing compressed bit vector with features which have altitude information.
+    succinct::rs_bit_vector altitudeAvailability(processor.GetAltitudeAvailability());
+    string const altitudeAvailabilityPath = my::JoinFoldersToPath(baseDir, "altitude_availability.bitvector");
```

Comment:
Also, this this in general wrong to create 3 files and directly catenate them, because succinct data structures require 8-bytes alignment. I insist that you should use a single writer instead of 3 files.